src/runtime/propensity-server.js

```diff
@@ -0,0 +1,262 @@
+            }
+            reject(new Error(errorMessage));
+          } else if (xhr.status < 200 || xhr.status > 300) {
+            const errorMessage = xhr.status + ':' + xhr.statusText;
+            reject(new Error(errorMessage));
+          } else {
+            const response = new Response();
+            response.status = xhr.status;
+            response.ok = true;
+            response.statusText = xhr.statusText;
+            xhr.bodyUsed = true;
+            response.responseText = xhr.responseText;
+            response.url = xhr.responseURL;
+            response.headers = this.getResponseHeader_(
+                xhr.getAllResponseHeaders());
+            resolve(response);
+          }
+        }
+      };
+      xhr.onerror = () => {
+        reject(new Error('Network failure'));
+      };
+      xhr.onabort = () => {
+        reject(new Error('Request aborted'));
+      };
+      if (init.method == 'POST') {
+        xhr.send(init.body);
+      } else {
+        xhr.send();
+      }
+    });
+  }
+
+  /**
+   * Creates header from XHR header
+   * @param {?string} responseHeaders
+   * @return {!Headers}
+   */
+  getResponseHeader_(responseHeaders) {
+    const headerMap = new Headers();
+    if (!responseHeaders) {
+      return headerMap;
+    }
+    const arr = responseHeaders.trim().split(/[/r/n]+/);
+    arr.forEach(function(line) {
+      const parts = line.split(': ');
+      const header = parts.shift();
+      const value = parts.join(': ');
+      headerMap[header] = value;
+    });
+    return headerMap;
+  }
+}
+
+/**
+ * Implements interface to Propensity server
+ */
+export class PropensityServer {
+  /**
+   * Page configuration is known when Propensity API
+   * is available, publication ID is therefore used
+   * in constructor for the server interface.
+   * @param {string} publicationId
+   */
+  constructor(win, publicationId) {
+    /** @private @const {!Window} */
+    this.win_ = win;
+    /** @private @const {string} */
+    this.publicationId_ = publicationId;
+    /** @private {?string} */
+    this.clientId_ = null;
+    /** @private {?string} */
+    this.product_ = null;
+    /** @private {boolean} */
+    this.userConsent_ = false;
+    /** @private @const {boolean}*/
+    this.TEST_SERVERS_ = true;
+    /** {!Object.<number, string>} */
+    const errorMap = {
+      404: 'Publisher not whitelisted',
+      403: 'Not sent from allowed origin',
+      400: 'Invalid request',
+      500: 'Server not available',
+    };
+    /** @private @const {!XhrInterface} */
+    this.xhrInterface_ = new XhrInterface(errorMap);
+  }
+
+  /**
+   * @return {string}
+   */
+  getUrl_() {
+    if (this.TEST_SERVERS_) {
+      return 'http://sohanirao.mtv.corp.google.com:8080';
+    } else {
+      return 'https://pubads.g.doubleclick.net/gampad/adx';
+    }
+  }
+
+  /**
+   * @return {?string}
+   */
+  getGads_() {
+    let gads = null;
+    const gadsmatch = this.win_.document.cookie.match(
```

Comment:
nit: can you put a comment describing what you're trying to match (for those of us who don't know how to read this off the top of our heads :)

Reply:
also - does gad stand for something?

Reply:
added